### DISCUSSION TITLE
Feature Request: Keyframe Library

### DISCUSSION THREAD

**Discussion by DWrightAnimation** [keiwando/toonsquid#338] [Ideas]
Whenever you make a symbol for the symbol library, you have the ability to create markers that allow you to select one of those poses. I recently discovered that when you change the timing from Hold to anything else, it goes to the next pose based on the timing. This would be a lifesaver, had it not been for the fact that if I wanted to go from the first pose to the pose after the second pose, I'd have to go through the timing that goes from marker 1 to marker 3. What I'd like is something that can keep my key poses saved, so I can use them like the markers, except instead of using the timing based on the symbol's internal timing, the project would make its own transition from one key pose to another. There are many poses that a smile can transition to, but currently, it can only transition to one pose at a time without it looking wrong.

**Comment by Cat-Butler** [comment]
Sorry, just trying to understand this.
Are you wanting to drop single static frames from a symbol (like a library of cut-and-paste images), and then draw the in-between frames manually?

Compared to the current version, which is more like the [pre-sculpted interchangeable mouth positions in stop-motion animation](https://www.vfxvoice.com/aardman-goes-back-to-stop-motion-basics-with-early-man/).
Currently your main timeline can go [smile],smile,smile,[surprise],surprise,[smile],smile,smile,[frown], frown. (Or something like [1],1,1,[3],3,[1],1,1,[2],2 from the symbol).
Actual keyframes marked with square brackets, automatic Hold in-between keyframes are shown without square brackets (it doesn't use the symbol timing as long as you stick to the constraints of one-named-frame-per-pose). You select/keyframe when the expression/pose changes in the main timeline.

You want:
[smile],{I'll draw an in-between here},{I'll draw an in-between here},{I'll draw an in-between here},[surprise]
where the keyframes in square brackets are dropped in from a symbol.

Do I understand this correctly?

I'm assuming you don't want some sort of auto-warp/smear in-between frames, which I can't imagine working well ([smile],{warp/smear},{warp/smear},[surprise])

**Reply by DWrightAnimation** [reply]
https://github.com/keiwando/toonsquid/assets/129681998/6a5c97cb-6cfc-480d-9336-4cbc826d039a

This is a recording of what currently is. The transition from Smile 2 to O is what I'm looking for in the transition from Smile 1 to O, but the transition currently goes from Smile 1 to Frown to O, which is inconvenient for making different mouth shapes ease into each other. Something as simple as singular mouth shapes to represent each sound would look too choppy for the style of animation I'm going for.

**Comment by Cat-Butler** [comment]
So, re-usable control point keyframes for Path objects - yes, I can see the usefulness of that - upvoted.

Would probably be a headache to implement if the symbol contained multiple layers, so it almost needs to be a separate tool or a different section in the Library.

**Comment by Cat-Butler** [comment]
A very clumsy workaround, which I'm sure you can see as well as I can: 
Keyframe those control point poses as several beginning frames of the main timeline, then copy-paste as needed. Far less productive than adding keyframes from a symbol's frame markers, or some similar equivalent.
